rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,379 @@
+
+Since Vrl allows for named parameters and optional parameters, the compiler
+will need to ensure the bytecode evaluates each parameter in a specific order
+(likely the order declared in the function). Bytecode will need to be emmitted
+for parameters that are not specified in the Vrl script to add a default value
+to the stack.
+
+### Flow analysis
+
+One of the expensive operations is path lookup.
+
+There may be certain situations where this can be improved by analyising the
+flow of data at compile time.
+
+For example, with this code:
+
+```coffee
+.thing[3].thang = 3
+log(.thing[3].thang)
+```
+
+the compiler can work out that the variable written to and read from is the
+same and can thus store that data in an array, any access can be made via a
+quick O(1) lookup using the index to the array.
+
+The flow analysis can become fairly complex.
+
+```coffee
+.message = parse_json!(.message)
+.message[3] = parse_json!(.message[2])
+log(.message[3].thing)
+```
+
+Functions like `parse_json` would need access to this array and hints as to
+what paths should be stored.
+
+A lot more thought needs to go into this before we can consider implementing
+it.
+
+#### Use a Bump Allocator?
+
+We could use a library such as [bumpalo](https://crates.io/crates/bumpalo).
+
+A bump allocator will allocate memory up front. This memory will then be used
+during the runtime. At the end of each transform the bump pointer is reset to
+the start - effectively deallocating all the memory used during the transform
+with very little cost.
+
+The downside to a bump allocator is we need to make sure sufficient memory is
+allocated up front.
+
+This is probably not ready yet until this PR for `BTreeMap` lands.
+https://github.com/rust-lang/rust/pull/77438
+
+
+### Optimization
+
+With the code as a single dimension array of Bytecode, it could be possible to
+scan the code for patterns and reorganise the Bytecode so it can run in a more
+optimal way.
+
+A lot more thought and research needs to go into this before we can consider
+implementing these changes.
+
+### Optimizing stdlib functions
+
+There is a small number of functions in the stdlib that follow a pattern of
+converting `Bytes` into a `Cow<str>` using `String::from_utf8_lossy`. The
+string is then allocated using `.into_owned()`.
+
+With some shifting around it should be possible to avoid this extra allocation,
+possibly at the cost of a slight reduction in code ergonomics.
+
+The functions that could possibly be optimised are:
+
+`contains`
+`ends_with`
+`starts_with`
+`join`
+`truncate`
+
+## Rationale
+
+Vrl is often identified as a bottleneck in performance tests. Being a key
+component in Vector any optimisation that can be done in Vrl will benefit
+Vector as a whole.
+
+## Drawbacks
+
+Downsides to moving Vrl to use reference counting:
+
+- Using `RefCell` does move the borrow checking to the runtime. Without compile
+  time checks the chances of a panic are much higher.
```

Comment:
It would be good to give some examples on what we should watch out for.

I would probably favour having a wrapper type that only exposes `try_borrow`, and having the error implement `ExpressionError`, so that we avoid panicking Vector in production.

Thinking about this, given that VRL functions return new values instead of mutated once, do we have a list of places we'd need to `borrow_mut`? If those are minimally required, we could keep that API internal to the compiler, to avoid exposing it to functions and thus significantly reducing the risk of errors.

Reply:
There are not many places where we do need to `borrow_mut` - mostly in the `Target` implementation. I think any panics would be indicative of a bug, so I think we should probably fail hard and make sure the bugs aren't there by testing rigorously.